# Lab | Pandas

In this lab, we will be working with the customer data from an insurance company, which can be found in the CSV file located at the following link: https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

The data includes information such as customer ID, state, gender, education, income, and other variables that can be used to perform various analyses.

Throughout the lab, we will be using the pandas library in Python to manipulate and analyze the data. Pandas is a powerful library that provides various data manipulation and analysis tools, including the ability to load and manipulate data from a variety of sources, including CSV files.

### Data Description

- Customer - Customer ID

- ST - State where customers live

- Gender - Gender of the customer

- Education - Background education of customers 

- Customer Lifetime Value - Customer lifetime value(CLV) is the total revenue the client will derive from their entire relationship with a customer. In other words, is the predicted or calculated value of a customer over their entire duration as a policyholder with the insurance company. It is an estimation of the net profit that the insurance company expects to generate from a customer throughout their relationship with the company. Customer Lifetime Value takes into account factors such as the duration of the customer's policy, premium payments, claim history, renewal likelihood, and potential additional services or products the customer may purchase. It helps insurers assess the long-term profitability and value associated with retaining a particular customer.

- Income - Customers income

- Monthly Premium Auto - Amount of money the customer pays on a monthly basis as a premium for their auto insurance coverage. It represents the recurring cost that the insured person must pay to maintain their insurance policy and receive coverage for potential damages, accidents, or other covered events related to their vehicle.

- Number of Open Complaints - Number of complaints the customer opened

- Policy Type - There are three type of policies in car insurance (Corporate Auto, Personal Auto, and Special Auto)

- Vehicle Class - Type of vehicle classes that customers have Two-Door Car, Four-Door Car SUV, Luxury SUV, Sports Car, and Luxury Car

- Total Claim Amount - the sum of all claims made by the customer. It represents the total monetary value of all approved claims for incidents such as accidents, theft, vandalism, or other covered events.


External Resources: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

## Challenge 1: Understanding the data

In this challenge, you will use pandas to explore a given dataset. Your task is to gain a deep understanding of the data by analyzing its characteristics, dimensions, and statistical properties.

- Identify the dimensions of the dataset by determining the number of rows and columns it contains.
- Determine the data types of each column and evaluate whether they are appropriate for the nature of the variable. You should also provide suggestions for fixing any incorrect data types.
- Identify the number of unique values for each column and determine which columns appear to be categorical. You should also describe the unique values of each categorical column and the range of values for numerical columns, and give your insights.
- Compute summary statistics such as mean, median, mode, standard deviation, and quartiles to understand the central tendency and distribution of the data for numerical columns. You should also provide your conclusions based on these summary statistics.
- Compute summary statistics for categorical columns and providing your conclusions based on these statistics.

In [2]:
# Import pandas

import pandas as pd
cars_polices = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv")
cars_polices

Customer          ST GENDER             Education  \
0     RB50392  Washington    NaN                Master   
1     QZ44356     Arizona      F              Bachelor   
2     AI49188      Nevada      F              Bachelor   
3     WW63253  California      M              Bachelor   
4     GA49547  Washington      M  High School or Below   
...       ...         ...    ...                   ...   
4003      NaN         NaN    NaN                   NaN   
4004      NaN         NaN    NaN                   NaN   
4005      NaN         NaN    NaN                   NaN   
4006      NaN         NaN    NaN                   NaN   
4007      NaN         NaN    NaN                   NaN   

     Customer Lifetime Value   Income  Monthly Premium Auto  \
0                        NaN      0.0                1000.0   
1                 697953.59%      0.0                  94.0   
2                1288743.17%  48767.0                 108.0   
3                 764586.18%      0.0                 106.0   
4                 536307.65%  36357.0                  68.0   
...                      ...      ...                   ...   
4003                     NaN      NaN                   NaN   
4004                     NaN      NaN                   NaN   
4005                     NaN      NaN                   NaN   
4006                     NaN      NaN                   NaN   
4007                     NaN      NaN                   NaN   

     Number of Open Complaints     Policy Type  Vehicle Class  \
0                       1/0/00   Personal Auto  Four-Door Car   
1                       1/0/00   Personal Auto  Four-Door Car   
2                       1/0/00   Personal Auto   Two-Door Car   
3                       1/0/00  Corporate Auto            SUV   
4                       1/0/00   Personal Auto  Four-Door Car   
...                        ...             ...            ...   
4003                       NaN             NaN            NaN   
4004                       NaN             NaN            NaN   
4005                       NaN             NaN            NaN   
4006                       NaN             NaN            NaN   
4007                       NaN             NaN            NaN   

      Total Claim Amount  
0               2.704934  
1            1131.464935  
2             566.472247  
3             529.881344  
4              17.269323  
...                  ...  
4003                 NaN  
4004                 NaN  
4005                 NaN  
4006                 NaN  
4007                 NaN  

[4008 rows x 11 columns]

In [3]:
# Identify the dimensions of the dataset by determining the number of rows and columns it contains.
cars_polices.shape

(4008, 11)

In [4]:
# Determine the data types of each column and evaluate whether they are appropriate for the nature of the variable.

cars_polices.dtypes

Customer                      object
ST                            object
GENDER                        object
Education                     object
Customer Lifetime Value       object
Income                       float64
Monthly Premium Auto         float64
Number of Open Complaints     object
Policy Type                   object
Vehicle Class                 object
Total Claim Amount           float64
dtype: object

In [5]:
# You should also provide suggestions for fixing any incorrect data types.
cars_polices.sample(20)

Customer          ST GENDER Education Customer Lifetime Value   Income  \
3864      NaN         NaN    NaN       NaN                     NaN      NaN   
539   LL62746      Oregon   Male    Doctor              897064.73%  12829.0   
2576      NaN         NaN    NaN       NaN                     NaN      NaN   
1080      NaN         NaN    NaN       NaN                     NaN      NaN   
2133      NaN         NaN    NaN       NaN                     NaN      NaN   
2142      NaN         NaN    NaN       NaN                     NaN      NaN   
921   YG10247      Oregon      M   College             2295189.20%  62396.0   
103   EV68375  California    NaN   College              433038.60%  60475.0   
1483      NaN         NaN    NaN       NaN                     NaN      NaN   
3141      NaN         NaN    NaN       NaN                     NaN      NaN   
2197      NaN         NaN    NaN       NaN                     NaN      NaN   
638   DM95829  California      M   College              252395.96%  16244.0   
794   HG33568     Arizona      F    Master              748263.95%  25666.0   
289   FE84989          WA   Male   College              503574.46%  72672.0   
1775      NaN         NaN    NaN       NaN                     NaN      NaN   
2615      NaN         NaN    NaN       NaN                     NaN      NaN   
1205      NaN         NaN    NaN       NaN                     NaN      NaN   
3133      NaN         NaN    NaN       NaN                     NaN      NaN   
1402      NaN         NaN    NaN       NaN                     NaN      NaN   
455   ID20929      Oregon      M  Bachelor              275694.17%  29926.0   

      Monthly Premium Auto Number of Open Complaints     Policy Type  \
3864                   NaN                       NaN             NaN   
539                  118.0                    1/0/00   Personal Auto   
2576                   NaN                       NaN             NaN   
1080                   NaN                       NaN             NaN   
2133                   NaN                       NaN             NaN   
2142                   NaN                       NaN             NaN   
921                   64.0                    1/0/00   Personal Auto   
103                  107.0                    1/0/00   Personal Auto   
1483                   NaN                       NaN             NaN   
3141                   NaN                       NaN             NaN   
2197                   NaN                       NaN             NaN   
638                   68.0                    1/0/00   Personal Auto   
794                   63.0                    1/2/00   Personal Auto   
289                   63.0                    1/0/00   Personal Auto   
1775                   NaN                       NaN             NaN   
2615                   NaN                       NaN             NaN   
1205                   NaN                       NaN             NaN   
3133                   NaN                       NaN             NaN   
1402                   NaN                       NaN             NaN   
455                   74.0                    1/0/00  Corporate Auto   

      Vehicle Class  Total Claim Amount  
3864            NaN                 NaN  
539             SUV          328.231432  
2576            NaN                 NaN  
1080            NaN                 NaN  
2133            NaN                 NaN  
2142            NaN                 NaN  
921   Four-Door Car          307.200000  
103   Four-Door Car          513.600000  
1483            NaN                 NaN  
3141            NaN                 NaN  
2197            NaN                 NaN  
638    Two-Door Car          623.223617  
794   Four-Door Car          270.002766  
289   Four-Door Car          259.361117  
1775            NaN                 NaN  
2615            NaN                 NaN  
1205            NaN                 NaN  
3133            NaN                 NaN  
1402            NaN               

We could observe some suggestions to fix datatypes:

1/ There are rows with all "NaN" fields (empty rows).
2/ There are categorical columns with weird number of unique values. For example, we have GENDER column with 5 unique values. We have to reduce these values to 3 at least (Female, Male and Nan)
3/ For the column "Number of open complaints" we have an object value, so we need to convert it to number values. It seems to show the date of these complaints, but for customers who have no complaints, it shows "1/0/0" which is incorrect.

In [6]:
# Identify the number of unique values for each column and determine which columns appear to be categorical.
cars_polices.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4008 entries, 0 to 4007
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Customer                   1071 non-null   object 
 1   ST                         1071 non-null   object 
 2   GENDER                     954 non-null    object 
 3   Education                  1071 non-null   object 
 4   Customer Lifetime Value    1068 non-null   object 
 5   Income                     1071 non-null   float64
 6   Monthly Premium Auto       1071 non-null   float64
 7   Number of Open Complaints  1071 non-null   object 
 8   Policy Type                1071 non-null   object 
 9   Vehicle Class              1071 non-null   object 
 10  Total Claim Amount         1071 non-null   float64
dtypes: float64(3), object(8)
memory usage: 344.6+ KB


In [7]:
cars_polices.columns.unique() # Unique values of columns

Index(['Customer', 'ST', 'GENDER', 'Education', 'Customer Lifetime Value',
       'Income', 'Monthly Premium Auto', 'Number of Open Complaints',
       'Policy Type', 'Vehicle Class', 'Total Claim Amount'],
      dtype='object')

In [8]:
cars_polices.nunique() # Unique values for each column

Customer                     1071
ST                              8
GENDER                          5
Education                       6
Customer Lifetime Value      1027
Income                        774
Monthly Premium Auto          132
Number of Open Complaints       6
Policy Type                     3
Vehicle Class                   6
Total Claim Amount            761
dtype: int64

In [9]:
# You should also describe the unique values of each categorical column and the range of values for numerical columns, and give your insights.

# Categorical columns
categorical_columns = cars_polices.select_dtypes(include='object').head()
categorical_columns

Customer          ST GENDER             Education Customer Lifetime Value  \
0  RB50392  Washington    NaN                Master                     NaN   
1  QZ44356     Arizona      F              Bachelor              697953.59%   
2  AI49188      Nevada      F              Bachelor             1288743.17%   
3  WW63253  California      M              Bachelor              764586.18%   
4  GA49547  Washington      M  High School or Below              536307.65%   

  Number of Open Complaints     Policy Type  Vehicle Class  
0                    1/0/00   Personal Auto  Four-Door Car  
1                    1/0/00   Personal Auto  Four-Door Car  
2                    1/0/00   Personal Auto   Two-Door Car  
3                    1/0/00  Corporate Auto            SUV  
4                    1/0/00   Personal Auto  Four-Door Car

In [10]:
# Number of unique values for categorical columns
cars_polices.select_dtypes(include='object').nunique()


Customer                     1071
ST                              8
GENDER                          5
Education                       6
Customer Lifetime Value      1027
Number of Open Complaints       6
Policy Type                     3
Vehicle Class                   6
dtype: int64

In [29]:
# Unique values for each categorical column
# Customer unique values
cars_polices.Customer.value_counts()


Customer
RB50392    1
HJ15383    1
AO74776    1
HQ82233    1
OL72737    1
          ..
RO26085    1
ES57969    1
JK55587    1
RN97635    1
MY31220    1
Name: count, Length: 1071, dtype: int64

In [31]:
# State (ST) unique values
cars_polices.ST.value_counts()

ST
Oregon        320
California    211
Arizona       186
Cali          120
Nevada         98
Washington     81
WA             30
AZ             25
Name: count, dtype: int64

In [12]:
# GENDER unique values
cars_polices.GENDER.value_counts()


GENDER
F         457
M         413
Male       39
female     28
Femal      17
Name: count, dtype: int64

In [13]:
# Education unique values
cars_polices.Education.value_counts()

Education
Bachelor                324
College                 313
High School or Below    296
Master                   94
Doctor                   37
Bachelors                 7
Name: count, dtype: int64

In [14]:
# "Customer Lifetime Value" unique values
cars_polices["Customer Lifetime Value"].value_counts()

Customer Lifetime Value
445811.34%     4
251459.20%     4
272535.64%     3
578018.22%     3
684615.03%     3
              ..
245357.08%     1
507566.27%     1
321497.94%     1
1227534.31%    1
899704.02%     1
Name: count, Length: 1027, dtype: int64

In [15]:
# "Number of Open Complaints" unique values
cars_polices["Number of Open Complaints"].value_counts()

Number of Open Complaints
1/0/00    830
1/1/00    138
1/2/00     50
1/3/00     34
1/4/00     13
1/5/00      6
Name: count, dtype: int64

In [16]:
# "Policy type" unique values
cars_polices["Policy Type"].value_counts()

Policy Type
Personal Auto     780
Corporate Auto    234
Special Auto       57
Name: count, dtype: int64

In [17]:
# "Vehicle class" unique values
cars_polices["Vehicle Class"].value_counts()

Vehicle Class
Four-Door Car    576
Two-Door Car     205
SUV              199
Sports Car        57
Luxury SUV        20
Luxury Car        14
Name: count, dtype: int64

In [22]:
#Numerical columns
numerical_columns = cars_polices.select_dtypes(include='number').head() # Numerical columns
numerical_columns

Income  Monthly Premium Auto  Total Claim Amount
0      0.0                1000.0            2.704934
1      0.0                  94.0         1131.464935
2  48767.0                 108.0          566.472247
3      0.0                 106.0          529.881344
4  36357.0                  68.0           17.269323

In [19]:
# Numbers of unique values for each number columns
cars_polices.select_dtypes(include='number').nunique()

Income                  774
Monthly Premium Auto    132
Total Claim Amount      761
dtype: int64

In [20]:
#Range of values for numerical columns
for col in cars_polices.select_dtypes(include=["int64", "float64"]).columns:
    print(f"Range of {col}: {cars_polices[col].max()}")

Range of Income: 99960.0
Range of Monthly Premium Auto: 35354.0
Range of Total Claim Amount: 2893.239678


In [ ]:
# INSIGHTS

CATEGORICAL COLUMNS INSIGHTS
    - The percentage of women over men is aproximately 10%
    - The level of studies of customer is mostly Bachelor and College
    - The most contracted policy is Personal Auto
    - The vehicle class most common is Four-Door car

NUMERICAL COLUMNS INSIGHTS
    - The range of total Claim Amount is $2893.

In [24]:
# Compute summary statistics such as mean, median, mode, standard deviation, and quartiles to understand the central tendency and distribution of the data for numerical columns. You should also provide your conclusions based on these summary statistics.
numerical_columns.describe()

Income  Monthly Premium Auto  Total Claim Amount
count      5.000000              5.000000            5.000000
mean   17024.800000            275.200000          449.558557
std    23721.470564            405.488841          466.816080
min        0.000000             68.000000            2.704934
25%        0.000000             94.000000           17.269323
50%        0.000000            106.000000          529.881344
75%    36357.000000            108.000000          566.472247
max    48767.000000           1000.000000         1131.464935

NUMERICAL COLUMNS STATISTICS


In [27]:
# Compute summary statistics for categorical columns and providing your conclusions based on these statistics.
def compute_summary_statistics(cars_polices, categorical_columns):
    for column in categorical_columns:
        print(f"Summary Statistics for {column}:\n")
        frequency_count = cars_polices[column].value_counts()
        percentage_distribution = cars_polices[column].value_counts(normalize=True) * 100
        
        # Display results
        print("Frequency Count:\n", frequency_count)
        print("\nPercentage Distribution:\n", percentage_distribution)
        print("\n" + "-"*50 + "\n")

# Run the function on your dataset
compute_summary_statistics(cars_polices, categorical_columns)

Summary Statistics for Customer:

Frequency Count:
 Customer
RB50392    1
HJ15383    1
AO74776    1
HQ82233    1
OL72737    1
          ..
RO26085    1
ES57969    1
JK55587    1
RN97635    1
MY31220    1
Name: count, Length: 1071, dtype: int64

Percentage Distribution:
 Customer
RB50392    0.093371
HJ15383    0.093371
AO74776    0.093371
HQ82233    0.093371
OL72737    0.093371
             ...   
RO26085    0.093371
ES57969    0.093371
JK55587    0.093371
RN97635    0.093371
MY31220    0.093371
Name: proportion, Length: 1071, dtype: float64

--------------------------------------------------

Summary Statistics for ST:

Frequency Count:
 ST
Oregon        320
California    211
Arizona       186
Cali          120
Nevada         98
Washington     81
WA             30
AZ             25
Name: count, dtype: int64

Percentage Distribution:
 ST
Oregon        29.878618
California    19.701214
Arizona       17.366947
Cali          11.204482
Nevada         9.150327
Washington     7.563025
WA     

## Challenge 2: analyzing the data

### Exercise 1

The marketing team wants to know the top 5 less common customer locations. Create a pandas Series object that contains the customer locations and their frequencies, and then retrieve the top 5 less common locations in ascending order.

In [57]:
# Top 5 less common customer locations
states_counts = cars_polices.ST.value_counts() # Value counts of states.

In [58]:
# Sort values
states_counts.sort_values()

ST
AZ             25
WA             30
Washington     81
Nevada         98
Cali          120
Arizona       186
California    211
Oregon        320
Name: count, dtype: int64

In [78]:
# 5 smallest numbers
states_counts.nsmallest(5)

ST
AZ             25
WA             30
Washington     81
Nevada         98
Cali          120
Name: count, dtype: int64

### Exercise 2

The sales team wants to know the total number of policies sold for each type of policy. Create a pandas Series object that contains the policy types and their total number of policies sold, and then retrieve the policy type with the highest number of policies sold.

*Hint:*
- *Using value_counts() method simplifies this analysis.*
- *Futhermore, there is a method that returns the index of the maximum value in a column or row.*


In [73]:
# "Policy type" unique values
policy_counts = cars_polices["Policy Type"].value_counts()
policy_counts

Policy Type
Personal Auto     780
Corporate Auto    234
Special Auto       57
Name: count, dtype: int64

In [77]:
# "Policy type" maximum value
# policy_counts.idxmax() > show index name with max value
# policy_counts.max() > show max value of the index
print(f"The policy with the maximum value is {policy_counts.idxmax()} with total of {policy_counts.max()}.")

The policy with the maximum value is Personal Auto with total of 780.


### Exercise 3

The sales team wants to know if customers with Personal Auto have a lower income than those with Corporate Auto. How does the average income compare between the two policy types?

- Use *loc* to create two dataframes: one containing only Personal Auto policies and one containing only Corporate Auto policies.
- Calculate the average income for each policy.
- Print the results.

In [83]:
cars_polices["Policy Type"].value_counts()

Policy Type
Personal Auto     780
Corporate Auto    234
Special Auto       57
Name: count, dtype: int64

In [98]:
# Create Personal Auto dataframe
# Use loc to select rows with policy type
personal_auto_df = cars_polices.loc[cars_polices["Policy Type"] == "Personal Auto"]
personal_auto_df

# BEST ANSWER - Dont use loc if you know the row value you want to show
personal_auto_df = cars_polices[cars_polices["Policy Type"] == "Personal Auto"]
personal_auto_df


Customer          ST GENDER             Education  \
0     RB50392  Washington    NaN                Master   
1     QZ44356     Arizona      F              Bachelor   
2     AI49188      Nevada      F              Bachelor   
4     GA49547  Washington      M  High School or Below   
5     OC83172      Oregon      F              Bachelor   
...       ...         ...    ...                   ...   
1065  XD85577        Cali      M              Bachelor   
1066  TM65736      Oregon      M                Master   
1067  VJ51327        Cali      F  High School or Below   
1068  GS98873     Arizona      F              Bachelor   
1070  MY31220  California      F               College   

     Customer Lifetime Value   Income  Monthly Premium Auto  \
0                        NaN      0.0                1000.0   
1                 697953.59%      0.0                  94.0   
2                1288743.17%  48767.0                 108.0   
4                 536307.65%  36357.0                  68.0   
5                 825629.78%  62902.0                  69.0   
...                      ...      ...                   ...   
1065              284226.69%  69417.0                  73.0   
1066              305955.03%  38644.0                  78.0   
1067             2031499.76%  63209.0                 102.0   
1068              323912.47%  16061.0                  88.0   
1070              899704.02%  54230.0                 112.0   

     Number of Open Complaints    Policy Type  Vehicle Class  \
0                       1/0/00  Personal Auto  Four-Door Car   
1                       1/0/00  Personal Auto  Four-Door Car   
2                       1/0/00  Personal Auto   Two-Door Car   
4                       1/0/00  Personal Auto  Four-Door Car   
5                       1/0/00  Personal Auto   Two-Door Car   
...                        ...            ...            ...   
1065                    1/1/00  Personal Auto  Four-Door Car   
1066                    1/1/00  Personal Auto  Four-Door Car   
1067                    1/2/00  Personal Auto            SUV   
1068                    1/0/00  Personal Auto  Four-Door Car   
1070                    1/0/00  Personal Auto   Two-Door Car   

      Total Claim Amount  
0               2.704934  
1            1131.464935  
2             566.472247  
4              17.269323  
5             159.383042  
...                  ...  
1065           30.874869  
1066          361.455219  
1067          207.320041  
1068          633.600000  
1070          537.600000  

[780 rows x 11 columns]

In [99]:
# Create Corporate Auto dataframe
# Use loc to select rows with policy type
corporate_auto_df = cars_polices.loc[cars_polices["Policy Type"] == "Corporate Auto"]
corporate_auto_df

# BEST ANSWER - Dont use loc if you know the row value you want to show
corporate_auto_df = cars_polices[cars_polices["Policy Type"] == "Corporate Auto"]
corporate_auto_df

Customer          ST GENDER             Education  \
3     WW63253  California      M              Bachelor   
6     XZ87318      Oregon      F               College   
7     CF85061     Arizona      M                Master   
8     DY87989      Oregon      M              Bachelor   
12    DP39365  California    NaN                Master   
...       ...         ...    ...                   ...   
1045  WZ31900      Oregon      F              Bachelor   
1050  BJ53923     Arizona      M  High School or Below   
1052  FB23788      Oregon      M  High School or Below   
1056  PI47776      Oregon      F              Bachelor   
1059  YG44474      Oregon      M               College   

     Customer Lifetime Value   Income  Monthly Premium Auto  \
3                 764586.18%      0.0                 106.0   
6                 538089.86%  55350.0                  67.0   
7                 721610.03%      0.0                 101.0   
8                2412750.40%  14072.0                  71.0   
12                879879.70%  77026.0                 110.0   
...                      ...      ...                   ...   
1045              864970.06%  94389.0                 107.0   
1050              260027.21%  51978.0                  66.0   
1052              882883.50%  86721.0                 111.0   
1056              253070.51%  89451.0                  63.0   
1059             1401472.13%  54193.0                 117.0   

     Number of Open Complaints     Policy Type  Vehicle Class  \
3                       1/0/00  Corporate Auto            SUV   
6                       1/0/00  Corporate Auto  Four-Door Car   
7                       1/0/00  Corporate Auto  Four-Door Car   
8                       1/0/00  Corporate Auto  Four-Door Car   
12                      1/2/00  Corporate Auto  Four-Door Car   
...                        ...             ...            ...   
1045                    1/0/00  Corporate Auto            SUV   
1050                    1/0/00  Corporate Auto  Four-Door Car   
1052                    1/0/00  Corporate Auto            SUV   
1056                    1/0/00  Corporate Auto  Four-Door Car   
1059                    1/0/00  Corporate Auto            SUV   

      Total Claim Amount  
3             529.881344  
6             321.600000  
7             363.029680  
8             511.200000  
12            472.029737  
...                  ...  
1045           85.063708  
1050          144.782152  
1052          532.800000  
1056           61.769564  
1059          720.752945  

[234 rows x 11 columns]

In [94]:
# Calculate the income of each policy type
income_personal_auto = personal_auto_df.Income.sum() # Income Personal Auto
print(f"The Personal Auto income is ${income_personal_auto}")

income_corporate_auto = corporate_auto_df.Income.sum() # Income Corporate Auto
print(f"The Corporate Auto income is ${income_corporate_auto}")

The Personal Auto income is $29780945.0
The Corporate Auto income is $9685333.0


In [ ]:
# Compute summary statistics for categorical columns and providing your conclusions based on these statistics.
def compute_summary_statistics(cars_polices, categorical_columns):
    for column in categorical_columns:
        print(f"Summary Statistics for {column}:\n")
        frequency_count = cars_polices[column].value_counts()
        percentage_distribution = cars_polices[column].value_counts(normalize=True) * 100
        
        # Display results
        print("Frequency Count:\n", frequency_count)
        print("\nPercentage Distribution:\n", percentage_distribution)
        print("\n" + "-"*50 + "\n")

# Run the function on your dataset
compute_summary_statistics(cars_polices, categorical_columns)

Summary Statistics for Customer:

Frequency Count:
 Customer
RB50392    1
HJ15383    1
AO74776    1
HQ82233    1
OL72737    1
          ..
RO26085    1
ES57969    1
JK55587    1
RN97635    1
MY31220    1
Name: count, Length: 1071, dtype: int64

Percentage Distribution:
 Customer
RB50392    0.093371
HJ15383    0.093371
AO74776    0.093371
HQ82233    0.093371
OL72737    0.093371
             ...   
RO26085    0.093371
ES57969    0.093371
JK55587    0.093371
RN97635    0.093371
MY31220    0.093371
Name: proportion, Length: 1071, dtype: float64

--------------------------------------------------

Summary Statistics for ST:

Frequency Count:
 ST
Oregon        320
California    211
Arizona       186
Cali          120
Nevada         98
Washington     81
WA             30
AZ             25
Name: count, dtype: int64

Percentage Distribution:
 ST
Oregon        29.878618
California    19.701214
Arizona       17.366947
Cali          11.204482
Nevada         9.150327
Washington     7.563025
WA     

In [96]:
# BONUS: If you want to show a dataframe with no "Personal Auto" and "Corporate Auto"
cars_polices[(cars_polices["Policy Type"] != "Personal Auto") & (cars_polices["Policy Type"] != "Corporate Auto")]

Customer          ST GENDER             Education  \
9     BQ94931      Oregon      F               College   
52    RB69909      Nevada    NaN  High School or Below   
56    AP67935  California    NaN               College   
127   GO36627          AZ    NaN               College   
131   FV21968      Nevada    NaN              Bachelor   
...       ...         ...    ...                   ...   
4003      NaN         NaN    NaN                   NaN   
4004      NaN         NaN    NaN                   NaN   
4005      NaN         NaN    NaN                   NaN   
4006      NaN         NaN    NaN                   NaN   
4007      NaN         NaN    NaN                   NaN   

     Customer Lifetime Value   Income  Monthly Premium Auto  \
9                 738817.81%  28812.0                  93.0   
52                798343.17%  36014.0                  69.0   
56                761951.58%  64749.0                  64.0   
127               383745.19%  21880.0                  97.0   
131               597397.69%  41611.0                  74.0   
...                      ...      ...                   ...   
4003                     NaN      NaN                   NaN   
4004                     NaN      NaN                   NaN   
4005                     NaN      NaN                   NaN   
4006                     NaN      NaN                   NaN   
4007                     NaN      NaN                   NaN   

     Number of Open Complaints   Policy Type  Vehicle Class  \
9                       1/0/00  Special Auto  Four-Door Car   
52                      1/3/00  Special Auto  Four-Door Car   
56                      1/0/00  Special Auto  Four-Door Car   
127                     1/0/00  Special Auto  Four-Door Car   
131                     1/0/00  Special Auto  Four-Door Car   
...                        ...           ...            ...   
4003                       NaN           NaN            NaN   
4004                       NaN           NaN            NaN   
4005                       NaN           NaN            NaN   
4006                       NaN           NaN            NaN   
4007                       NaN           NaN            NaN   

      Total Claim Amount  
9             425.527834  
52            173.956072  
56            302.565190  
127           424.077159  
131           113.801497  
...                  ...  
4003                 NaN  
4004                 NaN  
4005                 NaN  
4006                 NaN  
4007                 NaN  

[2994 rows x 11 columns]

### Bonus: Exercise 4


Your goal is to identify customers with a high policy claim amount.

Instructions:

- Review again the statistics for total claim amount to gain an understanding of the data.
- To identify potential areas for improving customer retention and profitability, we want to focus on customers with a high policy claim amount. Consider customers with a high policy claim amount to be those in the top 25% of the total claim amount. Create a pandas DataFrame object that contains information about customers with a policy claim amount greater than the 75th percentile.
- Use DataFrame methods to calculate summary statistics about the high policy claim amount data. 

*Note: When analyzing data, we often want to focus on certain groups of values to gain insights. Percentiles are a useful tool to help us define these groups. A percentile is a measure that tells us what percentage of values in a dataset are below a certain value. For example, the 75th percentile represents the value below which 75% of the data falls. Similarly, the 25th percentile represents the value below which 25% of the data falls. When we talk about the top 25%, we are referring to the values that fall above the 75th percentile, which represent the top quarter of the data. On the other hand, when we talk about the bottom 25%, we are referring to the values that fall below the 25th percentile, which represent the bottom quarter of the data. By focusing on these groups, we can identify patterns and trends that may be useful for making decisions and taking action.*

*Hint: look for a method that gives you the percentile or quantile 0.75 and 0.25 for a Pandas Series.*

*Hint 2: check `Boolean selection according to the values of a single column` in https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9*

In [ ]:
# Your code here